### PR TITLE
Power control with RTC

### PR DESCRIPTION
**Pull request by rodrigocelmer**
This is the second example from issue #3

### CODE REVIEW

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: Is it entering standby or stop mode? I'm confused, because in one place you say standby and in the other place you say stop.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: Maybe it would be better if you create defines for the RTC's passwords, one define for each password.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: This one is entering in standby mode. I'm sorry, I just forgot to fix the comments

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: I saw that there're defines for it in one of the uC's library

Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: Actually, I saw defines for flash's passwords, not for RTC

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: According to the RM when entering Standby Mode you don't need to set LPSDSR bit, because the voltage regulator is going to be powered off.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: According to the RM the RCC->CSR register is protected against write in reset state, so shouldn't you be setting the PWR_CR_DBP bit before line 54?

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: As RCC->CSR register is write protected, shouldn't you set the PWR_DBP bit before set RCC_CSR_LSION?
And should'n you select the LSI as RTC clock before turning it on?

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c:57]: Nice! It is better with defines instead of aleatory hexadecimal numbers kkk

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: Why not 1Hz? And why not just RTC->PRER = 36?

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: You could comment here that you're making this operation for debouncing.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: Comment that this is how much time the uC will stay in Standby Mode

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: Fix the comment to RTC_WPR instead of RTC_WRP

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c:24]: Where are you using PA14 and PA13? If you're not using, remove the comments.

**Review by gustavoscherer — CHANGES_REQUESTED** [review]
I've made some comments in the lines that I've encountered something different from the RM, please talk tome in case of any doubt.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: Only LSEON, LSEBYP, RTCSEL, LSEDRV and RTCEN bits in RCC_CSR are write-protected. And I'm only resetting RTC for testing, it's not really necessary in this case

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: Same as the comment above.

> And should'n you select the LSI as RTC clock before turning it on?

I don't remember seeing anything about it in the RM, but makes sense

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: First thing, it should be **RTC_PRER_PREDIV_A**, not RTC_PRER_PREDIV_S. 
I had some difficulties to understand the Prescaler part in RTC, but I'm using the 7-bit asynchronous Prescaler for the RTC (it looks like to me that the synchronous one is not used for the wakeup event, and, for what I could understand, the asynchronous save more energy than the synchronous). So I couldn't get a 1 Hz output for the asynchronous Prescaler (289 Hz min)

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c:24]: They're used by the ST-link, for programming and debugging. It's more like a reminder of that, to not try to use them for other functions, and I've put this comment because you or Vini asked to in the EXTI example

Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c:24]: Ok!

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c]: The RM is pretty confusing about this part, because it doesn't say explicit that it's only these bits that are write protected. 

image

In this image it just say RCC CSR, it doesn't specify.

But I took a look in the TDR code and there we just set the DBP bit before write into the RTC's domain registers, so it's ok what you're doing.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c]: It's ok, you can turn it on before select it in RTC, thinking about it makes more sense then the other way.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c:53]: If you can write in the other register of RCC CSR without set the DBP bit, it's ok!

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: Here it should be DBGMCU_CR_DBG_STANDBY, because you're entering StandBy mode.

**Review by gustavoscherer — CHANGES_REQUESTED** [review]
Just the DBG bit is wrong.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c]: You can use both PREDIV_A and PREDIV_S for the auto wakeup, but if you want short timeout periods (up to 32s) you don't need to select the ck_spre as wakeup clock, you can use RTCCLOCK divided by 2, 4, 8 or 16. When using the both PREDIV_A and PREDIV_S you should set PREDIV_A to a high value, in order to minimize consumption, but it doesn't say if one prescaler save more energy than the other.
Anyway, it's ok to use 1kHz now, but I'm no sure about the maximum wakeup time we can get with this wakeup clock, because of the maximum value you can set in the RTC_WUPR register. So maybe for greater times you're gonna have to increase the prescalers.
Nice job until now!

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: since those enables are in the same register, I'll enable all of them together lol

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: I've thought it was the same bit and didn't check the RM. My bad

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: So, since I'm setting f_apre to 1 kHz and f_spre (1 Hz), what exactly is happening? Here's where I had some problems to understand all this clocking matter on RTC. I can't remember now why I'm setting them both, but it worked. My RTC clock input is 37 kHz (HSI).

Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c]: Both PREDIV_A and PREDIV_S have default value: PREDIV_A = 0x007F and PREDIV_S = 0x00FF. So if you want to use only the PREDIV_A, you have to set the PREDIV_S = 0x0000. Read the page 647 from the RM to understand better how the prescaler work, but basically, when using ck_spre as wakeup clock, the value of ck_spre will be the RTC clock (LSI in this case) divided by PREDIV_A and then divided by PREDIV_S, or RTC clock / [(PREDIV_A +1)*(PREDIV_S+1)].

And you're using LSI right?

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c]: It seems ok tome, but make sure you don't have to enable them in an order.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c:52]: Just make sure this register don't need to be set in an order.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: Try set the prescaler like this: RTC->PRER = 0x00240000. With this you're setting the PREDIV_A = 36 and PREDIV_S = 0.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c]: It's ok!

**Review by gustavoscherer — CHANGES_REQUESTED** [review]
Almost done!

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: Yes! LSI!

I'll check that again.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: But `RTC->PRER = (36 << RTC_PRER_PREDIV_A_Pos)` is equal to `RTC->PRER = 0x00240000`, which means PREDIV_A = 36 and PREDIV_S = 0.

I didn't understand your point on doing this

Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c]: It became a bit confusing! I thought that with this operation you wouldn't be setting the PREDIV_S = 0x0000, I'm not very good with shift operations. Sorry about this!

Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: Ahh, ok. I'm gonna try it

Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: I've got the same result by doing either `RTC->PRER = (36 << RTC_PRER_PREDIV_A_Pos)` or `RTC->PRER = 0x00240000`.

I hadn't understood your objective of doing this before.

Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c]: ok them!

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [inline PWR/PWR_RTC/Core/Src/main.c]: It's 1kHz right?

**Review by gustavoscherer — CHANGES_REQUESTED** [review]
Just fix the comment.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c:52]: It's ok enable all together!

Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c:52]: I took a quick look again on the RM (in RTC and RCC chapter) and couldn't find anything about this. But, since I'm doing `while(!(RCC->CSR & RCC_CSR_LSIRDY)); //poll until LSI is stable` after turning LSI on and enabling it as RTC clock source, I think it's ok to leave it as it is

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: no, I changed it. I would comment when I finished reading about the clock.

I've set PREDIV_A to 36 and PREDIV_S to 999, so I get a 1 Hz input for the RTC

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c:52]: Yeah, it's ok!

File: PWR/PWR_RTC/Core/Src/main.c
Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c]: Yes, but in PREDIV_A you have 1kHz, not 100Hz

**Comment by rodrigocelmer** [comment]
There'ya go

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: right... I calculated first to be 100 Hz, but PREDIV_A would be to high

Comment (gustavoscherer) [reply PWR/PWR_RTC/Core/Src/main.c]: Fix this and request the review again, please.

**Comment by rodrigocelmer** [comment]
@gustavoscherer I already requested for you to review it again, so I can't do it again. I fixed the comment on commit 8620258

**Review by gustavoscherer — APPROVED** [review]
I think it's everything ok now! I'm approving this code, let's wait for @Vinimuller review before merging.

Nice work, man! These functions using RTC and low power modes are very complex.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (Vinimuller) [inline PWR/PWR_RTC/Core/Src/main.c]: What are those operations between line 53 and 56?

File: PWR/PWR_RTC/Core/Src/main.c
Comment (Vinimuller) [inline PWR/PWR_RTC/Core/Src/main.c]: didn't understand what bits you are referring to on the comment

**Comment by rodrigocelmer** [comment]
@Vinimuller just a reminder :)

File: PWR/PWR_RTC/Core/Src/main.c
Comment (Vinimuller) [inline PWR/PWR_RTC/Core/Src/main.c:44]: Do we need to set ULP bit here?

File: PWR/PWR_RTC/Core/Src/main.c
Comment (Vinimuller) [reply PWR/PWR_RTC/Core/Src/main.c]: Ok, I got that you are reseting the RTC domain in order to configure it later and RM says that this bit should be set and cleared by sw. But my question is that it seems kinda useless that while after you set and reset the bit. Does RM indicates to do it? Will the code actually waits for something there?

File: PWR/PWR_RTC/Core/Src/main.c
Comment (Vinimuller) [inline PWR/PWR_RTC/Core/Src/main.c]: According to RM, if we need to use both prescalers, it is preferable to set the async prescaler to the highest value possible because of energy consumption. 
image
I would rather set async prescaler to 125 and the sync to 296

File: PWR/PWR_RTC/Core/Src/main.c
Comment (Vinimuller) [inline PWR/PWR_RTC/Core/Src/main.c:73]: As I read the code and tried to follow the RM steps to check if everything is ok, I found a bit difficult because it is in a different order. All the steps are being correctly followed but what I mean is that you disable the WUT as RM states on **Programming WakeUp Timer** but then you start the process of **Initialisation and configuration**. I would rather follow the initialisation steps, set the prescaler values, and then deal with the WUT

File: PWR/PWR_RTC/Core/Src/main.c
Comment (Vinimuller) [inline PWR/PWR_RTC/Core/Src/main.c:74]: In the examples of the RM they usually lock RTC access after configuring it (step 7):
image
Did you find any information on that?

**Review by Vinimuller — CHANGES_REQUESTED** [review]

File: PWR/PWR_RTC/Core/Src/main.c
Comment (Vinimuller) [inline PWR/PWR_RTC/Core/Src/main.c:64]: RM says that the reset state of this bit is zero. As you reseted RTC registers on line 53, I guess it's not needed to reset this bit again

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: I don't remember seeing anything about them in the RM, probably I put them there by myself, just to be sure that the bit is actually set and then reset. And, again, I'm just resetting the RTC so I'm sure that I'm writing to the registers and nothing more. I could remove it since both of you had questions about it.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c]: Actually, to all of them. Those three bits in CSR are write-protected and I forgot it once or twice while working on it

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c:44]: **Stop mode** chapter says: "_To minimize the consumption In Stop mode, VREFINT, the BOR, PVD, and temperature sensor can be switched off before entering Stop mode. This functionality is controlled by the ULP bit in the PWR_CR register._"

**Standby mode** chapter says: "_The Standby mode allows to achieve the lowest power consumption. It is based on the Cortex®-M0+ Deepsleep mode, **with the voltage regulator disabled**._" But it doesn't say if you need to set ULP bit or not. To avoid any doubts, I set it.

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c:73]: I'm very confused.

I didn't get your point. By doing as you said, we would be doing the opposite from what the RM says to do (prescalers first, then WUTF). And which initialization and configurations steps are you talking about?

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c:74]: Sounds good, doesn't work.

I used these so-called lock codes, other lock codes I've found on an ST forum and I cleared PWR_CR_DBP bit. I used all of them, one by one and even all together, and I could write to RTC registers still (upside-down smiling emoji)

File: PWR/PWR_RTC/Core/Src/main.c
Comment (rodrigocelmer) [reply PWR/PWR_RTC/Core/Src/main.c:64]: But it also says to "_clear WUTE in RTC_CR to disable the wakeup timer_" before setting the prescalers in "programming the wake-up timer"

File: PWR/PWR_RTC/Core/Src/main.c
Comment (Vinimuller) [inline PWR/PWR_RTC/Core/Src/main.c]: fix "standy mode" on comment

File: PWR/PWR_RTC/Core/Src/main.c
Comment (Vinimuller) [inline PWR/PWR_RTC/Core/Src/main.c:86]: Just add to your comment that it enables RTC wakeup timer

File: PWR/PWR_RTC/Core/Src/main.c
Comment (Vinimuller) [inline PWR/PWR_RTC/Core/Src/main.c]: standy mode here too

**Review by Vinimuller — CHANGES_REQUESTED** [review]
Just some comments fix. Nice work!

**Review by Vinimuller — APPROVED** [review]
Apruvedi

**Pull request merged**
